app/lib/shared/storage.dart

```diff
@@ -16,3 +28,118 @@ Future<Bucket> getOrCreateBucket(Storage storage, String name) async {
   }
   return storage.bucket(name);
 }
+
+Future deleteFromBucket(Bucket bucket, String objectName) async {
+  try {
+    await bucket.delete(objectName);
+  } on DetailedApiRequestError catch (e) {
+    if (e.status != 404) {
+      rethrow;
+    }
+  }
+}
+
+/// Uploads content from [openStream] to the [bucket] as [objectName].
+Future uploadWithRetry(Bucket bucket, String objectName, int length,
+    Stream<List<int>> openStream()) async {
+  await retryAsync(
+    () async {
+      final sink = bucket.write(objectName,
+          length: length, contentType: contentType(objectName));
+      await sink.addStream(openStream());
+      await sink.close();
+    },
+    description: 'Upload to $objectName',
+    shouldRetryOnError: (e) {
+      if (e is DetailedApiRequestError) {
+        return e.status == 502 || e.status == 503;
+      }
+      return false;
+    },
+    sleep: Duration(seconds: 10),
+  );
+}
+
+/// Uploads content from [bytes] to the [bucket] as [objectName].
+Future uploadBytesWithRetry(
+        Bucket bucket, String objectName, List<int> bytes) =>
+    uploadWithRetry(bucket, objectName, bytes.length,
+        () => new Stream.fromIterable([bytes]));
+
+/// Utility class to access versioned data file that follows the pattern:
+/// "/path-prefix/runtime-version.ext". The extension must use .gz as the
+/// storage uses transparent gzip on the data stream.
+class VersionedDataStorage {
+  final Bucket _bucket;
+  final String _prefix;
+  final String _extension;
+
+  VersionedDataStorage(Bucket bucket, String prefix, String extension)
+      : _bucket = bucket,
+        _prefix = prefix,
+        _extension = extension {
+    assert(extension.endsWith('.gz'));
+  }
+
+  /// Whether the storage bucket has a data file for the current runtime version.
+  /// TODO: decide whether we should re-generate the file after a certain age
+  Future<bool> hasCurrentData() async {
+    try {
+      final info = await _bucket.info(_objectName());
+      return info != null;
+    } catch (_) {
+      return false;
+    }
+  }
+
+  /// Upload the current data file to the storage bucket.
+  Future uploadDataFile(File file) async {
```

Comment:
let's raise the level a bit here...

Specifically, the `VersionedDataStorage` class should either be:
 A) a class that you store versioned binary files in (upload from file, download to file)
 B) a class that you store versioned binary bytes in (upload from `List<int>`, download to `List<int>`) 
 C) a class that you store versioned JSON blobs in (upload from JSON, download to JSON)

I don't really care which one we choose... Obviously, you have a plan to reuse it multiple places :)
But I care that we build abstractions that does something consistent.

Reply:
We can restrict the interface to JSON blob store, the only downside is that the upload of the SDK's `pub-data.json` would need to be first deserialized and then serialized again before uploading. Because that is a one-time event, the performance is not critical.

I don't have an immediate reuse for `VersionedDataStorage`. The static assets of the generated dartdoc archives are a bit similar, but that use case is about putting the files in a versioned subdirectory, and there is no case for JSON blobs there.

I'll update the interface to handle only JSON blobs.